### PR TITLE
mixer: add support for variable data copy

### PR DESCRIPTION
**Pull request by tlauda**
Adds support to mixer component for variable
data size copy.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

File: src/audio/mixer.c [hunk 1]
```diff
@@ -39,42 +39,48 @@
...
 
-#define trace_mixer(__e)	trace_event(TRACE_CLASS_MIXER, __e)
-#define tracev_mixer(__e)	tracev_event(TRACE_CLASS_MIXER, __e)
-#define trace_mixer_error(__e)	trace_error(TRACE_CLASS_MIXER, __e)
+#define trace_mixer(__e, ...) \
+	trace_event(TRACE_CLASS_MIXER, __e, ##__VA_ARGS__)
+#define tracev_mixer(__e, ...) \
+	tracev_event(TRACE_CLASS_MIXER, __e, ##__VA_ARGS__)
+#define trace_mixer_error(__e, ...) \
+	trace_error(TRACE_CLASS_MIXER, __e, ##__VA_ARGS__)
```
Comment (dbaluta) [inline src/audio/mixer.c]: This should be in a separate patch, with an explanation of why it is needed.

File: src/audio/mixer.c [hunk 2]
```diff
@@ -250,41 +261,44 @@ static int mixer_copy(struct comp_dev *dev)
...
+
 	/* mix streams */
-	md->mix_func(dev, sink, sources, i, dev->frames);
+	md->mix_func(dev, sink, sources, i, frames);
 
-	/* update source buffer pointers for overflow */
+	/* update source buffer pointers */
 	for (i = --num_mix_sources; i >= 0; i--)
-		comp_update_buffer_consume(sources[i], md->period_bytes);
+		comp_update_buffer_consume(sources[i], source_bytes);
```
Comment (ranj063) [inline src/audio/mixer.c]: @tlauda what is the assumption here in using only the source_bytes calculated for source[0] and consuming the same number of bytes from all the sources?

File: src/audio/mixer.c [hunk 3]
```diff
@@ -140,16 +146,16 @@ static int mixer_params(struct comp_dev *dev)
...
+	if (period_bytes == 0) {
 		trace_mixer_error("mixer_params() error: period_bytes = 0");
 		return -EINVAL;
 	}
 
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
 
 	/* set downstream buffer size */
-	ret = buffer_set_size(sink, md->period_bytes * config->periods_sink);
+	ret = buffer_set_size(sink, period_bytes * config->periods_sink);
```
Comment (ranj063) [inline src/audio/mixer.c]: @tlauda should this be done in prepare() using the params from the sink?

File: src/audio/mixer.c (hunk 2 above)
Comment (tlauda) [reply src/audio/mixer.c]: @ranj063 Mixer expects the same format from all the sources so we can base our calculations just on one of the sources. We are calculating the min bytes, so the consumption value would be valid for all of them.

File: src/audio/mixer.c (hunk 3 above)
Comment (tlauda) [reply src/audio/mixer.c]: @ranj063 Mixer doesn't change the format, so it makes no difference.

File: src/audio/mixer.c (hunk 2 above)
Comment (ranj063) [reply src/audio/mixer.c]: @tlauda then can you please add a comment here stating this.
Also, you need to add a check to make sure this is indeed the case.

File: src/audio/mixer.c [hunk 4]
```diff
@@ -253,41 +261,47 @@ static int mixer_copy(struct comp_dev *dev)
 	if (num_mix_sources == 0)
 		return 0;
 
-	/* make sure no sources have underruns */
-	for (i = 0; i < num_mix_sources; i++) {
+	/* check for overrun */
+	if (sink->free == 0) {
```
Comment (ranj063) [inline src/audio/mixer.c]: @tlauda this can be if (!sink->free). But why dont we check is sink->free >= source_bytes here?

Comment (tlauda) [reply src/audio/mixer.c]: @ranj063 The intention of the patch is to remove the need of limiting copy to period_size. We should copy as much data as we possible can.

Comment (ranj063) [reply src/audio/mixer.c]: @tlauda that intent needs to be in a comment or in the commit message :)

**Comment by lgirdwood** [comment]
@tlauda CI UT failing
